src/Microsoft.Azure.ServiceBus/Core/Batch.cs

```diff
@@ -0,0 +1,144 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.ServiceBus.Core
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using Microsoft.Azure.Amqp;
+    using Microsoft.Azure.Amqp.Framing;
+    using Microsoft.Azure.ServiceBus.Amqp;
+    using Microsoft.Azure.ServiceBus.Diagnostics;
+
+    [DebuggerDisplay("{" + nameof(DebuggerDisplay) + ",nq}")]
+    public class Batch : IDisposable
+    {
+        internal readonly ulong maximumBatchSize;
+        private readonly Func<Message, Task<Message>> pluginsCallback;
+        private AmqpMessage firstMessage;
+        private readonly List<Data> datas;
+        private AmqpMessage result;
+        private (string messageId, string sessionId, string partitionKey, string viaPartitionKey) originalMessageData;
+
+        /// <summary>
+        /// Construct a new batch with a maximum batch size and outgoing plugins callback.
+        /// <remarks>
+        /// To construct a batch at run-time, use <see cref="MessageSender"/>, <see cref="QueueClient"/>, or <see cref="TopicClient"/>.
+        /// Use this constructor for testing and custom implementations.
+        /// </remarks>
+        /// </summary>
+        /// <param name="maximumBatchSize">Maximum batch size allowed for batch.</param>
+        /// <param name="pluginsCallback">Plugins callback to invoke on outgoing messages regisered with batch.</param>
+        public Batch(ulong maximumBatchSize, Func<Message, Task<Message>> pluginsCallback)
```

Comment:
>public Batch [](start = 8, length = 12)

I think having this constructor will cause more confusion. We already have a way to create a batch on the sender. 

For testing as of now, we could make this internal. 
For customers who need to test this, they could mock a Sender (in the future version) and then create a batch accordingly. 

Having this constructor exposed like this will definitely create confusion to customers which is just explained in remarks (which is not the most accessible comment).

I would think especially the second param - callback is going to be even more confusing for a beginner user. 
Lets make this internal for now.